README.md
# RainRadar

[เว็บไซต์](http://radar.pkn.sh) | [ซอร์สโค้ด](https://github.com/pknme/rainradar) | [ประวัติการเปลี่ยนแปลง](https://github.com/pknme/rainradar/releases)

RainRadar เป็นเว็บไซต์สำหรับดูภาพเรดาร์ตรวจอากาศทั่วประเทศไทยและข้อมูลเพิ่มเติมอื่นๆ เข้าใช้งานได้ผ่าน [radar.pkn.sh](http://radar.pkn.sh)

---

## คุณสมบัติ

- รวมภาพเรดาร์ตรวจอากาศทั่วประเทศไทยจาก
  - [กรมอุตุนิยมวิทยา](http://weather.tmd.go.th/)
  - [กรมฝนหลวงและการบินเกษตร](http://122.154.75.14/RRMThaiGov/RadarApp/RadarMainRoyalRain.php)
  - [สำนักการระบายน้ำ กรุงเทพมหานคร](http://weather.bangkok.go.th/radar/)
- ข้อมูลการจราจรจากทวิตเตอร์
- ข้อมูลพยากรณ์อากาศทั่วประเทศประจำวัน
- ข้อมูลคุณภาพอากาศ (AQI)
- ข้อมูลรายงานและพยากรณ์ฝนบริเวณกรุงเทพและปริมณฑล
- ข้อมูลสถิติฝนในพื้นที่กรุงเทพมหานคร
- ลิงก์ไปยังระบบตรวจวัดต่างๆ ของสำนักการระบายน้ำ กรุงเทพมหานคร

## ขอบคุณ

ขอบคุณโครงการเหล่านี้ที่ผมใช้ในการพัฒนา RainRadar

- [Visual Studio Code](https://code.visualstudio.com/)
- [Materalize](http://materializecss.com/)
- [images.weserv.nl](https://images.weserv.nl/)
- [jQuery](https://jquery.com)
- [GitHub](https://github.com)
- [Cloudflare](https://www.cloudflare.com/)

---

## LICENSE

ซอร์สโค้ดของ RainRadar เผยแพร่ภายใต้[สัญญาอนุญาต MIT](https://github.com/pknme/rainradar/blob/master/LICENSE.md)


CHANGELOG.md
# บันทึกการเปลี่ยนแปลงของ RainRadar

## RainRadar 8.6.0 (19 ก.พ. 2563)

- เพิ่มข้อมูลสภาพอากาศในหน้าแรก (beta)
- เพิ่มเรดาร์ชัยนาท
- ปรับปรุงการดูภาพเคลื่อนไหว
- ปรับปรุงเมนู
- ปรับปรุงหน้า AQI เพิ่มข้อมูลในพื้นที่กรุงเทพและภาคใต้
- ปรับปรุงลิงก์ AQI ในหน้าแรก
- รองรับประกาศพิเศษในหน้าแรก
- ปรับปรุงสคริปต์พยากรณ์อากาศและพื้นที่เสี่ยงภัยฝนตกหนัก
- เปลี่ยนโดเมนเป็น radar.pkn.sh
- ลบโค้ดและไฟล์ที่ไม่ได้ใช้ออก

## RainRadar 8.5.1 (21 ก.ย. 2562)

- แยกประกาศกรมอุตุนิยมวิทยาออกมาเป็นหน้าเดี่ยว
- แก้ไขลิงก์ภาพเคลื่อนไหวของเรดาร์ประจวบคีรีขันธ์ (หัวหิน)
- เพิ่มข้อมูลเวลาในภาพเรดาร์ในด้านล่างของหน้าเว็บ
- แก้บั๊กเมื่อไม่สามารถโหลดรูปพยากรณ์อากาศ
- แก้ไขข้อผิดพลาดอื่นๆ อีกเล็กน้อย

## RainRadar 8.5.0 (13 ก.ย. 2562)

- เพิ่มภาพเคลื่อนไหวให้เรดาร์ลำพูน, ประจวบคีรีขันธ์ (หัวหิน) และภูเก็ต
- นำเรดาร์สิงห์บุรี (อินทร์บุรี), ระยอง, ชุมพร และกระบี่ออก
- แก้ไขลิงก์เรดาร์ลำพูนและภูเก็ต
- ปรับปรุงการเนื้อหาในหน้าแรก
- ปรับปรุงหน้า AQI ใหม่
- ปรับปรุงเนื้อหาบางส่วน
- เปลี่ยนโดเมนเซิร์ฟเวอร์ภาพเรดาร์และไฟล์ฟอนต์
- ย้ายหน้าค้นหาไปไว้บน Notion
- เพิ่มแบบสอบถามโดยใช้ Emojicom
- แก้ไขข้อผิดพลาดอื่นๆ อีกเล็กน้อย

## RainRadar 8.4.2 (21 ธ.ค. 2561)

- เพิ่มสถานีข้อมูลคุณภาพอากาศ (AQI)
  - ราษฎร์บูรณะ, กทม.
  - บางนา, กทม.
  - สะพานควาย, กทม.
  - สมุทรสาคร
  - พระนครศรีอยุธยา
- ลบสถานีข้อมูลคุณภาพอากาศ (AQI) เนื่องจากไม่มีข้อมูลอัปเดต
  - นครราชสีมา
  - ภูเก็ต

## RainRadar 8.4.1 (12 ธ.ค. 2561)

- เพิ่มเรดาร์นครสวรรค์ (ตาคลี)

## RainRadar 8.4.0 (1 พ.ย. 2561)

- เพิ่มเรดาร์ระยอง
- รวมเรดาร์ภาคตะวันออกไว้ในเมนูภาคกลาง
- แสดงเมนูทั้งหมดเมื่อเปิดในหน้าจอขนาดเล็ก
- หัวข้อเว็บไซต์จะเปลี่ยนตามข้อมูลที่ดู
- เพิ่มแท็กลิงก์ในเมนูภาพเรดาร์
- เพิ่มลิงก์ไปยังข้อมูลเกี่ยวกับตัวบล็อกโฆษณา
- แก้ไขลิงก์ในหน้าช่วยเหลือ
- แก้ไข URL ไม่อัปเดตตามข้อมูลในบางกรณี
- ปรับปรุงฟอนต์ในส่วนประกาศกรมอุตุนิยมวิทยา

## RainRadar 8.3.3 (13 ก.ย. 2561)

- เพิ่มเรดาร์อุบลราชธานี (120 ก.ม.)

## RainRadar 8.3.2 (11 ก.ย. 2561)

- เพิ่มและแก้ไขชื่อสถานีข้อมูลคุณภาพอากาศ (AQI)
  - เพิ่มใหม่
    - แหลมฉบัง, ชลบุรี
    - ลำพูน
    - ท่าแพ, เชียงใหม่
    - ศูนย์ราชการ, เชียงใหม่
    - แม่ริม, เชียงใหม่
    - แม่ฮ่องสอน
    - น่าน
    - นราธิวาส
  - เปลี่ยนชื่อ
    - คลองหลวง, ปทุมธานี เปลี่ยนชื่อเป็น ม.กรุงเทพ, ปทุมธานี
  - นำออก
    - เชียงใหม่
    - พะเยา
- แก้ไขข้อมูลเครดิตในหน้าเกี่ยวกับ

## RainRadar 8.3.1 (10 ก.ย. 2561)

- เพิ่มสถานีข้อมูลคุณภาพอากาศ (AQI)
  - ศาลาแดง, กทม.
  - วงเวียนใหญ่, กทม.
  - โชคชัย 4, กทม.
  - คลองหลวง, ปทุมธานี
  - ชลบุรี
  - พะเยา
  - นครราชสีมา
- ปรับปรุง UI หน้าข้อมูลคุณภาพอากาศ (AQI)
- แก้ไข URL หน้าแรกไม่ถูกต้องหลังปิดหน้าต่างข้อมูล

## RainRadar 8.3.0 (10 ก.ย. 2561)

- เพิ่มประกาศล่าสุดจากกรมอุตุนิยมวิทยาในหน้าแรก
- รองรับการแสดงข้อมูลสถานะหากมีเหตุขัดข้อง
- ย้ายข้อมูลคุณภาพอากาศ (AQI) จากหน้าแรกไปเป็นหน้าแยกต่างหาก
- เปลี่ยนลิงก์ไปยังหน้าสถานะเป็น [radarstatus.openbase.co](https://radarstatus.openbase.co)
- เปลี่ยนลิงก์หน้าช่วยเหลือ
- โหลดฟอนต์จาก CDN, เพิ่มฟอนต์ตัวหนา
- ปรับปรุง UI และอื่นๆ อีกเล็กน้อย

## RainRadar 8.2.0 (9 ส.ค. 2561)

- เปลี่ยนเซิร์ฟเวอร์ภาพเรดาร์จาก [storage1-cdn.pakin.me](https://storage1-cdn.pakin.me) เป็น [imgcdn.pakin.me](https://imgcdn.pakin.me) ดาวน์โหลดภาพเรดาร์ได้เร็วขึ้น
- เพิ่มเรดาร์
  - ภูเก็ต
  - ประจวบคีรีขันธ์ (หัวหิน)
- เปลี่ยนชื่อเรดาร์ลำปาง (เกาะคา) เป็น แพร่ (ร้องกวาง) ตามเว็บไซต์กรมอุตุนิยมวิทยา
- เพิ่มข้อความแจ้งเตือนค่าสีและหน่วยที่ใช้วัดของแต่ละเรดาร์
- ปรับปรุงเนื้อหาในหน้าช่วยเหลือ
- ปรับปรุงการส่งข้อมูลสถิติการใช้งาน

## RainRadar 8.1.0 (26 ก.ค. 2561)

- เปลี่ยนโดเมนจาก [radar.pknme.com](http://radar.pknme.com) เป็น [radar.openbase.co](https://radar.openbase.co/)
- เพิ่มแผนที่พื้นที่เสี่ยงภัยฝนตกหนักถึงหนักมากในหน้าแรก
- ปรับปรุงการโหลดภาพพยากรณ์อากาศในหน้าแรก

## RainRadar 8.0.1 (13 ก.ค. 2561)

- เพิ่มลิงก์ไปยังหน้าสถานะ

## RainRadar 8.0.0 (15 มิ.ย. 2561)

- เพิ่มลิงก์ไปยังหน้าวิธีใช้งานเบื้องต้น, บันทึกการเปลี่ยนแปลง รูปแบบใหม่
- เพิ่มเรดาร์
  - ลำปาง (เกาะคา)
  - สุราษฎร์ธานี
  - กระบี่ (ภาพเคลื่อนไหว)
- ลบเรดาร์
  - หนองบอน
  - หนองบอน (ภาพเคลื่อนไหว)
  - ระยอง
  - ระยอง (ภาพเคลื่อนไหว)
  - ประจวบคีรีขันธ์ (หัวหิน)
  - ภูเก็ต
  - ภูเก็ต (ภาพเคลื่อนไหว)
- เปลี่ยน ID ของเรดาร์บางตัว
- เพิ่มข้อความแจ้งเตือนให้ปิดตัวบล็อกโฆษณาในหน้าข้อมูลการจราจร
- ปรับปรุงการโหลดภาพเรดาร์เคลื่อนไหว
- ปรับปรุงรูปแบบเมื่อเปิดผ่าน HTTP
- เปลี่ยนเซิร์ฟเวอร์สำหรับโหลดภาพเรดาร์
- เพิ่ม Google Analytics
- ปรับปรุงรูปแบบกล่องป๊อปอัป
- ปรับปรุง UI และอื่นๆ อีกเล็กน้อย

## RainRadar 7.1.1 (22 มี.ค. 2561)

- ลบเรดาร์หนองบอน
- ปรับปรุงข้อมูลคุณภาพอากาศ
- ปรับปรุงสคริปต์เล็กน้อย
- แก้ไขลิงก์ไฟล์ CSS ในหน้า 404

## RainRadar 7.1.0 (28 ก.พ. 2561)

- ปรับปรุง[ภาพรวมเรดาร์ทั้งประเทศ](https://radar.pknme.com/#all)ให้โหลดภาพในหน้าเว็บเหมือนเรดาร์อื่นๆ
- เพิ่มลิงก์ไปยังหน้ารวมเรดาร์ภาพเคลื่อนไหว
- เพิ่มไฟล์บึนทึกการเปลี่ยนแปลง
- ใช้ CDN กับไฟล์ฟอนต์
- ปรับปรุงการโหลดไฟล์
- การปรับปรุงอื่นๆ อีกเล็กน้อย

## RainRadar 7.0.0 (22 ก.พ. 2561)

ในอัปเดตนี้ RainRadar เริ่มบังคับการใช้งานผ่าน HTTPS พร้อมเพิ่มข้อมูลคุณภาพอากาศจาก [aqicn.org](http://aqicn.org) และปรับปรุงเพิ่มประสิทธิภาพการโหลดข้อมูลต่างๆ

### ทั่วไป

- เปลี่ยนที่อยู่เว็บไซต์จาก [radar.pakin.me](https://radar.pakin.me) เป็น [radar.pknme.com](https://radar.pknme.com)
- เพิ่มข้อตกลงในการใช้งานและนโยบายความเป็นส่วนตัว

### HTTPS

- บังคับใช้งานผ่าน HTTPS
- เพิ่มหน้าสำหรับดูภาพเรดาร์แบบเคลื่อนไหวผ่าน HTTP ([http://nossl.radar.pknme.com](http://nossl.radar.pknme.com))
- เขียนสคริปต์สำหรับโหลดภาพผ่าน HTTPS ใหม่ทั้งหมด
- เพิ่มประสิทธิภาพการโหลดข้อมูลต่างๆ

### เรดาร์

- เพิ่มเรดาร์หนองบอน
- เพิ่มเรดาร์หนองบอน (ภาพเคลื่อนไหว)
- เพิ่มเรดาร์หนองบอน (ภาพละเอียดสูง)
- นำเรดาร์หนองจอก (ทิศทางลม) ออก
- นำเรดาร์หนองแขม (ทิศทางลม) ออก
- เพิ่มการรองรับเมื่อเรดาร์มีปัญหาหรือปิดใช้งานชั่วคราว
- เพิ่มรูปภาพสำรองเมื่อโหลดภาพเรดาร์ไม่ได้

### หน้าแรก

- เพิ่มข้อมูลคุณภาพอากาศจาก [aqicn.org](http://aqicn.org)
- ปรับปรุงการโหลดรูปภาพพยากรณ์อากาศ
- เพิ่มลิงก์ไปยังแบบฟอร์มติดต่อ

### หน้าช่วยเหลือ

- ปรับปรุงเนื้อหาในหน้าช่วยเหลือ
- เพิ่มลิงก์ไปยังแบบฟอร์มติดต่อ

### อื่นๆ

- อัปเดตไฟล์ robots.txt
- เพิ่มไฟล์ฟอนต์ในตัว
- เพิ่มหน้า 404
- ปรับปรุง UI เล็กน้อย

main.nossl.js
var hash = window.location.hash,
  img = $(".radar-img"),
  ldg = $(".loading"),
  stxt = $(".status-text"),
  tw = $(".twitter-block"),
  loadtw = $(".loadtwitter"),
  mdcnt = $(".modal .modal-content"),
  togglessl = $(".togglessl");

$(document).ready(function() {

  $(".loadradar").click(function() {
    var data = $(this).data(),
      radarName = $(this).text(),
      radarId = $(this).attr("id"),
      d = new Date(),
      date = d.getFullYear() + "" + (d.getMonth() + 1) + "" + d.getDate() + "" + d.getHours(),
      minute5 = d.getMinutes() - (d.getMinutes() % 5),
      minute10 = d.getMinutes() - (d.getMinutes() % 10),
      time5 = date + minute5,
      time20 = date + minute10;
    if (minute10 > 20) {
      time20 = date + "20";
    }
    window.history.replaceState({}, "", "/#" + radarId);
    $('.nav-extended').css("top", "0");
    tw.hide();
    img.removeAttr("src");
    img.show();
    img.attr("src", data.imgprefix + data.img + "?cv84t=" + time5);
    if (data.type == "info") {
      stxt.html("<h5>" + data.title + "</h5>");
    } else {
      stxt.html("<h5>เรดาร์" + radarName + " (ภาพเคลื่อนไหว)</h5>");
    }
    ldg.show();
    img.on('load', function() {
      ldg.hide();
      if (data.type == "info") {
        stxt.html("<h5>" + data.title + "</h5>ข้อมูลจาก" + data.src);
      } else {
        stxt.html("<h5>เรดาร์" + radarName + " (ภาพเคลื่อนไหว)</h5>ภาพเรดาร์จาก" + data.src);
      }
    });
    img.on('error', function() {
      ldg.hide();
      img.hide();
      if (data.type == "info") {
        stxt.html("<h5>" + data.title + "</h5>ไม่สามารถโหลดข้อมูลได้<br>กรุณาลองใหม่อีกครั้งในอีก 5-10 นาที หรือเลือกดูภาพเรดาร์แทน");
      } else if (location.protocol == 'https:' && data.imgani == "yes") {
        stxt.load("/content/imgani.html");
      } else {
        stxt.html("<h5>เรดาร์" + radarName + "</h5>ไม่สามารถโหลดภาพเรดาร์<br>กรุณาลองใหม่อีกครั้งในอีก 5-10 นาที หรือเลือกดูเรดาร์อื่น");
      }
    });
  });
  
  $('.dropdown-menu').dropdown({
    inDuration: 325,
    outDuration: 200,
    constrain_width: false,
    hover: false,
    gutter: 0,
    belowOrigin: true,
    alignment: 'left'
  });

  // Load image by url
  $(hash).click();

});



main.js
var hash = window.location.hash,
  img = $('.radar-img'),
  ldg = $('.loading'),
  stxt = $('.status-text'),
  tw = $('.twitter-block'),
  loadtw = $('.loadtwitter'),
  mdcnt = $('.modal .modal-content'),
  //togglessl = $(".togglessl"),
  twnotice = $('.traffic-notice'),
  unitnotice = $('.radar-unit-notice'),
  weatherBlock = $('.weather-block');

if (hash == '') {
  window.history.replaceState({}, '', '/#home');
  stxt.load('../content/home.html?v=8.6.0-c2');
  weatherBlock.show();
  gtag('config', 'UA-78233854-2', {
    page_path: '/#home',
  });
}

$(document).ready(function() {
  twnotice.hide();

  $('.loadradar').click(function(e) {
    e.preventDefault();
    var data = $(this).data(),
      radarName = data.title,
      aniRadarName = data.anititle,
      radarId = $(this).attr('id'),
      d = new Date(),
      date =
        d.getFullYear() +
        '' +
        (d.getMonth() + 1) +
        '' +
        d.getDate() +
        '' +
        d.getHours(),
      minute5 = d.getMinutes() - (d.getMinutes() % 5),
      minute10 = d.getMinutes() - (d.getMinutes() % 10),
      time5 = date + minute5,
      time20 = date + minute10;
    if (minute10 > 20) {
      time20 = date + '20';
    }
    window.history.replaceState({}, '', '/#' + radarId);
    gtag('config', 'UA-78233854-2', {
      page_path: '/#' + radarId,
    });
    $('.nav-extended').css('top', '0');
    weatherBlock.hide();
    tw.hide();
    twnotice.hide();
    img.removeAttr('src');
    img.show();
    if (data.cache == '20') {
      img.attr(
        'src',
        'https://rainradarcontent.pkn.sh/i/' +
          time20 +
          '/' +
          data.img +
          '?rainradar=8.6.0-c2',
      );
      //img.attr("src", "https://images.weserv.nl?url=" + data.img + "?rainradar=8.6.0-c2" + time20);
    } else {
      img.attr(
        'src',
        'https://rainradarcontent.pkn.sh/i/' +
          time5 +
          '/' +
          data.img +
          '?rainradar=8.6.0-c2',
      );
      //img.attr("src", "https://images.weserv.nl?url=" + data.img + "?rainradar=8.6.0-c2" + time5);
      unitnotice.show();
    }
    if (data.type == 'info') {
      stxt.html('<h5>' + data.title + '</h5>ข้อมูลจาก' + data.src);
      document.title = data.title + ' | RainRadar';
    } else {
      stxt.html('<h5>เรดาร์' + radarName + '</h5>ภาพเรดาร์จาก' + data.src);
      document.title = 'เรดาร์' + radarName + ' | RainRadar';
    }
    ldg.show();
    img.on('load', function() {
      ldg.hide();
      if (data.imganiold === 'yes' && location.protocol === 'https:') {
        stxt.html(
          '<h5>RainRadar เปลี่ยนรูปแบบการโหลดภาพเคลื่อนไหว</h5>กรุณาเลือกเรดาร์ที่ต้องการดูใหม่ครั้ง',
        );
      }
    });
    img.on('error', function() {
      ldg.hide();
      unitnotice.hide();
      img.hide();
      if (data.type === 'info') {
        stxt.html(
          '<h5>' +
            data.title +
            '</h5>ไม่สามารถโหลดข้อมูลได้<br>กรุณาลองใหม่อีกครั้งในอีก 5-10 นาที หรือเลือกดูภาพเรดาร์แทน',
        );
      } else {
        stxt.html(
          '<h5>เรดาร์' +
            radarName +
            '</h5>ไม่สามารถโหลดภาพเรดาร์<br>กรุณาลองใหม่อีกครั้งในอีก 5-10 นาที หรือเลือกดูเรดาร์อื่น',
        );
      }
    });
  });

  $('.modal-trigger').click(function() {
    modalHash = window.location.hash;
    var data = $(this).data();
    $('.modal').attr('alt', data.identity);
    window.history.replaceState({}, '', '/#' + data.identity);
    gtag('config', 'UA-78233854-2', {
      page_path: '/#' + data.identity,
    });
    mdcnt.load(data.url, function(response, status) {
      if (status == 'error') {
        mdcnt.html('ไม่สามารถโหลดข้อมูล');
      }
    });
  });

  $('#modal').modal({
    opacity: 0.6,
    in_duration: 425,
    out_duration: 250,
    startingTop: '0%',
    endingTop: '8%',
    complete: function() {
      mdcnt.html('กำลังโหลดข้อมูล กรุณารอสักครู่');
      if (modalHash.match(/about|ddslinks|radarclosed|enableweather/)) {
        window.history.replaceState({}, '', '/#home');
        stxt.load('../content/home.html?v=8.6.0-c2');
        weatherBlock.show();
        gtag('config', 'UA-78233854-2', {
          page_path: '/#home',
        });
      } else if (modalHash === '#home') {
        window.history.replaceState({}, '', '#home');
        gtag('config', 'UA-78233854-2', {
          page_path: '/#home',
        });
      } else {
        window.history.replaceState({}, '', modalHash);
        gtag('config', 'UA-78233854-2', {
          page_path: '/' + modalHash,
        });
      }
    },
  });

  $('.dropdown-menu').dropdown({
    inDuration: 325,
    outDuration: 200,
    constrain_width: false,
    hover: false,
    gutter: 0,
    belowOrigin: true,
    alignment: 'left',
  });

  loadtw.click(function() {
    document.title = 'ข้อมูลการจราจร | RainRadar';
    window.history.replaceState({}, '', '/#traffic');
    gtag('config', 'UA-78233854-2', {
      page_path: '/#traffic',
    });
    window.scrollTo(0, 0);
    $('.nav-extended').css('top', '0');
    if (loadtw.data('click') == 0) {
      img.removeAttr('src');
      unitnotice.hide();
      img.hide();
      ldg.hide();
      weatherBlock.hide();
      stxt.html('<h5>ข้อมูลการจราจรจากทวิตเตอร์</h5>');
      ldg.show();
      twnotice.show();

      window.twttr = (function(d, s, id) {
        var js,
          fjs = d.getElementsByTagName(s)[0],
          t = window.twttr || {};
        if (d.getElementById(id)) return t;
        js = d.createElement(s);
        js.id = id;
        js.src = 'https://platform.twitter.com/widgets.js';
        fjs.parentNode.insertBefore(js, fjs);

        t._e = [];
        t.ready = function(f) {
          t._e.push(f);
        };

        return t;
      })(document, 'script', 'twitter-wjs');

      twttr.ready(function(twttr) {
        twttr.events.bind('rendered', function(event) {
          loadtw.data('click', 1);
          ldg.hide();
          tw.show();
          stxt.html(
            '<h5>ข้อมูลการจราจรจากทวิตเตอร์</h5>อัปเดตอัตโนมัติ เมื่อมีข้อมูลใหม่จะแสดงทันที',
          );
        });
      });
    } else {
      img.removeAttr('src');
      unitnotice.hide();
      img.hide();
      weatherBlock.hide();
      stxt.html(
        '<h5>ข้อมูลการจราจรจากทวิตเตอร์</h5>อัปเดตอัตโนมัติ เมื่อมีข้อมูลใหม่จะแสดงทันที',
      );
      ldg.hide();
      tw.show();
    }
  });

  $('#home').click(function(e) {
    e.preventDefault();
    window.history.replaceState({}, '', '/#home');
    gtag('config', 'UA-78233854-2', {
      page_path: '/#home',
    });
    $('.nav-extended').css('top', '0');
    img.removeAttr('src');
    stxt.html('');
    ldg.hide();
    unitnotice.hide();
    img.hide();
    tw.hide();
    twnotice.hide();
    stxt.load('../content/home.html?v=8.6.0-c2');
    weatherBlock.show();
  });

  // Load image by url
  $(hash).click();

  // Open modal by url
  if (hash.match(/about|ddslinks|radarclosed|enableweather/)) {
    $('#modal').modal('open');
  }

  // Get alert message
  $.getJSON(
    'https://api.sheety.co/ff305574-1482-4f55-b34b-5b987844a6d9',
    function(data) {
      if (data[0].display != true) {
        console.log('No alert message 🎉');
      } else if (data[0].simple === true) {
        content = '<p><b class="red-text">' + data[0].title + '</b>';
        content += ' ' + data[0].content + ' ';
        content +=
          '<a href="' +
          data[0].link +
          '" target="' +
          data[0].link_target +
          '">' +
          data[0].link_text +
          '</a></p>';
        $(content).appendTo('.alert-message-block');
        $('.alert-message-block')
          .css('display', 'block')
          .addClass('simple-alert');
      } else {
        content =
          '<h5 class="alert-message-title block-header flow-text">' +
          data[0].title +
          '</h5>';
        content +=
          '<p class="alert-message-content">' + data[0].content + '</p>';
        content +=
          '<a href="' +
          data[0].link +
          '" target="' +
          data[0].link_target +
          '" class="alert-message-link btn waves-effect ' +
          data[0].link_color +
          '">' +
          data[0].link_text +
          '</a>';
        $(content).appendTo('.alert-message-block');
        $('.alert-message-block')
          .css('display', 'block')
          .addClass(data[0].block_color + ' card-panel white-text');
      }
    },
  );

  // Hide navbar when scroll down
  var didScroll;
  var lastScrollTop = 0;
  var delta = 5;
  var navbarHeight = $('.nav-extended').outerHeight();
  var wrapperHeight = $('.nav-wrapper').outerHeight();

  $(window).scroll(function(event) {
    didScroll = true;
  });

  setInterval(function() {
    if (didScroll) {
      hasScrolled();
      didScroll = false;
    }
  }, 200);

  function hasScrolled() {
    var st = $(this).scrollTop();

    // Make sure they scroll more than delta
    if (Math.abs(lastScrollTop - st) <= delta) return;

    // If they scrolled down and are past the navbar, add class .nav-up.
    // This is necessary so you never see what is "behind" the navbar.
    if (st > lastScrollTop && st > navbarHeight) {
      // Scroll Down
      $('.nav-extended').css('top', -wrapperHeight);
    } else {
      // Scroll Up
      if (st + $(window).height() < $(document).height()) {
        $('.nav-extended').css('top', '0');
      }
    }

    lastScrollTop = st;
  }
});

// Weather

var deniedText =
  'คุณปฏิเสธสิทธิ์ในการเข้าถึงตำแหน่ง <a class="white-text" href="https://go.pkn.sh/r5vc7GjF" target="_blank">คลิกที่นี่เพื่อดูวิธีอนุญาตสิทธิ์อีกครั้ง</a>';

function handlePermission() {
  if (navigator.geolocation) {
    navigator.permissions.query({ name: 'geolocation' }).then(function(result) {
      console.log('geo: ' + result.state);
      if (result.state == 'granted') {
        getLocation();
      } else if (result.state == 'prompt') {
        // Do nothing
      } else if (result.state == 'denied') {
        $('.weather-placeholder').html(deniedText);
      }
      result.onchange = function() {
        console.log('geo: ' + result.state);
      };
    });
  } else {
    weatherBlock.remove();
    console.log('this browser not support geolocation api.');
  }
}

handlePermission();

function enableWeather() {
  $('#modal, .modal-overlay').hide();
  getLocation();
}

function getLocation() {
  $('.weather-placeholder').remove();
  $('.weather-loading').show();
  if (navigator.geolocation) {
    navigator.geolocation.getCurrentPosition(loadWeather, geoError, {
      timeout: 10000,
    });
  } else {
    $('#modal').modal('close');
    alert('เบราว์เซอร์ของคุณไม่รองรับการระบุตำแหน่ง');
    $('.weather-loading').html('เบราว์เซอร์ของคุณไม่รองรับการระบุตำแหน่ง');
  }
}

function loadWeather(position) {
  var lat = position.coords.latitude;
  var long = position.coords.longitude;
  $('#modal').modal('close');
  fetch(
    'https://api.pkn.sh/weather/data/2.5/weather?lat=' +
      lat +
      '&lon=' +
      long +
      '&units=metric&APPID=8aa6ba4495812c798eebe9e6ac17ecc9&lang=th',
  )
    .then(function(resp) {
      return resp.json();
    }) // Convert data to json
    .then(function(data) {
      drawWeather(data); // Call drawWeather
    })
    .catch(function() {
      // catch any errors
    });
}

function geoError(error) {
  $('#modal').modal('close');
  if (error.code == error.PERMISSION_DENIED) {
    alert('คุณปฏิเสธสิทธิ์ในการเข้าถึงตำแหน่ง คุณสมบัตินี้จะไม่ทำงาน');
    $('.weather-loading').html(deniedText);
  } else {
    alert('เกิดข้อผิดพลาดขณะระบุตำแหน่ง กรุณารีเฟรชหน้าเว็บแล้วลองอีกครั้ง');
    $('.weather-loading').html(
      'เกิดข้อผิดพลาดขณะระบุตำแหน่ง กรุณารีเฟรชหน้าเว็บแล้วลองอีกครั้ง',
    );
  }
}

function drawWeather(d) {
  var weatherTimestamp = dayjs.unix(d.dt).format('HH:mm');
  var sunrise = dayjs.unix(d.sys.sunrise).format('HH:mm');
  var sunset = dayjs.unix(d.sys.sunset).format('HH:mm');
  $('.description').html(d.weather[0].description);
  $('.temp').html(d.main.temp + '&deg;C');
  // $ ('.temp-min').html ('ต่ำสุด ' + d.main.temp_min + '&deg;C');
  // $ ('.temp-max').html ('สูงสุด ' + d.main.temp_max + '&deg;C');
  $('.weather-container .block-header').html('สภาพอากาศใน ' + d.name);
  $('.weather-icon').attr(
    'src',
    'https://static.pkn.sh/rainradar/weather-icons/' +
      d.weather[0].icon +
      '.svg',
  );
  $('.sunrise').html(sunrise + ' น.');
  $('.sunset').html(sunset + ' น.');
  $('.pressure').html(d.main.pressure + ' hPa');
  $('.wind-speed').html(d.wind.speed + ' เมตร/วินาที');
  $('.clouds').html(d.clouds.all + '%');
  $('.humidity').html(d.main.humidity + '%');
  $('.weather-timestamp').html('อัปเดตล่าสุด ' + weatherTimestamp + ' น.');
  $('.weather-link').attr('href', 'https://openweathermap.org/city/' + d.id);
  $('.weather-loading').hide();
  $('.weather-content').show();
}
